File: src/main/java/com/dh/ondot/schedule/api/response/ScheduleCreateResponse.java
package com.dh.ondot.schedule.api.response;

import com.dh.ondot.core.util.TimeUtils;
import com.dh.ondot.schedule.domain.Schedule;

import java.time.Duration;
import java.time.LocalDateTime;

public record ScheduleCreateResponse(
        Long scheduleId,
        int estimateTime,
        LocalDateTime preparationTriggeredAt,
        LocalDateTime departureTriggeredAt,
        LocalDateTime createdAt
) {
    public static ScheduleCreateResponse of(Schedule schedule) {
        LocalDateTime prepAt = TimeUtils.toSeoulDateTime(
                schedule.getPreparationAlarm().getTriggeredAt()
        );
        LocalDateTime deptAt = TimeUtils.toSeoulDateTime(
                schedule.getDepartureAlarm().getTriggeredAt()
        );
        LocalDateTime created = TimeUtils.toSeoulDateTime(schedule.getCreatedAt());
        int estimateTime = Math.toIntExact(
                Duration.between(deptAt, TimeUtils.toSeoulDateTime(schedule.getAppointmentAt()))
                        .toMinutes()
        );

        return new ScheduleCreateResponse(
                schedule.getId(),
                estimateTime,
                prepAt,
                deptAt,
                created
        );
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/api/response/ScheduleCreateResponse.kt
package com.dh.ondot.schedule.api.response

import com.dh.ondot.core.util.TimeUtils
import com.dh.ondot.schedule.domain.Schedule
import java.time.Duration
import java.time.LocalDateTime

data class ScheduleCreateResponse(
    val scheduleId: Long?,
    val estimateTime: Int,
    val preparationTriggeredAt: LocalDateTime?,
    val departureTriggeredAt: LocalDateTime?,
    val createdAt: LocalDateTime?,
) {
    companion object {
        @JvmStatic
        fun of(schedule: Schedule): ScheduleCreateResponse {
            val prepAt = TimeUtils.toSeoulDateTime(
                schedule.preparationAlarm!!.triggeredAt
            )
            val deptAt = TimeUtils.toSeoulDateTime(
                schedule.departureAlarm!!.triggeredAt
            )
            val created = TimeUtils.toSeoulDateTime(schedule.createdAt)
            val estimateTime = Math.toIntExact(
                Duration.between(deptAt, TimeUtils.toSeoulDateTime(schedule.appointmentAt))
                    .toMinutes()
            )

            return ScheduleCreateResponse(
                scheduleId = schedule.id,
                estimateTime = estimateTime,
                preparationTriggeredAt = prepAt,
                departureTriggeredAt = deptAt,
                createdAt = created,
            )
        }
    }
}
